rdmd_test.d

```diff
@@ -310,6 +310,16 @@ void runTests()
         assert(res.status == -SIGSEGV, format("%s", res));
     }
 
+    /* https://issues.dlang.org/show_bug.cgi?id=11997- rdmd should search it's
+    binary path for the compiler */
+    string localDMD = buildPath(tempDir(), baseName(compiler));
+    std.file.write(localDMD, "empty shell");
+    scope(exit) std.file.remove(localDMD);
+
+    res = execute(rdmdApp ~ [modelSwitch, "--force", "--chatty", "--eval=writeln(`Compiler found.`);"]);
+    assert(res.status == 1, res.output);
+    assert(res.output.canFind(`spawn ["` ~ localDMD ~ `",`));
```

Comment:
Added the test: it would be nice if we didn't expect it to fail, but that would mean copying a whole dmd install over, which would be too much of a PITA.  I'm not sure if the `--force` flag is necessary, didn't need it myself, but all other invocations of `--eval` in this test file use it, so I stuck it here too.

Reply:
This new test fails on NixOS because the build runs in a sandboxed environment and tempDir is the same dir as the build dir -> /tmp/nix-build-dtools-2.078.0.drv-5.
Which means that dtools is in /tmp/nix-build-dtools-2.078.0.drv-5/dtools and I also need to place the dmd source at /tmp/nix-build-dtools-2.078.0.drv-5/dmd which is the same path localDMD wants to be written to. :-)
I fixed it for now by moving the dmd dir before running the tests but since posix.mak needs the dmd dir I need to run the tests manually and can not use the posix.mak.
Wouldn't it be nicer if the dmd dir needs to be placed inside the dtools dir? You are not in control of anything above dtools dir in general.

Here is the error message:

std.file.FileException@std/file.d(626): /tmp/nix-build-dtools-2.078.0.drv-5/dmd: Is a directory
----------------
/nix/store/hg5rh9llch87i5153rfwn4f8nzgg08h2-dmdBuild-2.078.0/include/d2/std/file.d:263 @trusted bool std.file.cenforce!(bool).cenforce(bool, const(char)[], const(char)*, immutable(char)[], ulong) [0x4aef4a]
??:? @trusted void std.file.writeImpl(const(char)[], const(char)*, const(void[]), bool) [0x4bfd4c]
/nix/store/hg5rh9llch87i5153rfwn4f8nzgg08h2-dmdBuild-2.078.0/include/d2/std/file.d:534 @safe void std.file.write!(immutable(char)[]).write(immutable(char)[], const(void[])) [0x4a6e24]
rdmd_test.d:330 void rdmd_test.runTests() [0x48d201]
rdmd_test.d:79 _Dmain [0x489696]

Reply:
It seems nuts that your tempDir is the one where all your source packages are unpacked, surely there's a better way.

>Wouldn't it be nicer if the dmd dir needs to be placed inside the dtools dir? 

Not sure what you mean, these tests simply assume that the dmd compiler is installed somewhere in the system path and that there's no random file/directory named dmd lying around your tempDir, where it runs the tests.

Reply:
> It seems nuts that your tempDir is the one where all your source packages are unpacked, surely there's a better way.

It's not in my hand to change that. That's how the package manager Nix works for building packages. You are only allowed to write inside your build directory because you are not allowed to change anything else on the system this way. The package has only access to that very directory, this is important to prevent any side effects while building and maintain a reproducible build.

> Not sure what you mean, these tests simply assume that the dmd compiler is installed somewhere in the system path and that there's no random file/directory named dmd lying around your tempDir, where it runs the tests.

Currently dmd's source directory needs to be placed in the parent directory of the source directory of the tools. I suggest to expect the dmd source inside the tools directory. This is a much cleaner way because you don't know your environment and ../dmd might be an old dmd needed by the user or whatever which might lead to strange errors.
This would also fix my issue with Nix and prevent my current workaround.

Reply:
> Currently dmd's source directory needs to be placed in the parent directory of the source directory of the tools. I suggest to expect the dmd source inside the tools directory.

Not sure how that's related to this discussion, but potentially having a copy of every component inside the directory of every component doesn't sound better than each component expecting its siblings to be in its parent directory.

@ThomasMader Feel free to submit a PR which changes the `rdmdApp` and `localDMD` paths to a more unique subdirectory of `tempDir()`.

Reply:
>  Which means that dtools is in /tmp/nix-build-dtools-2.078.0.drv-5/dtools and I also need to place the dmd source at /tmp/nix-build-dtools-2.078.0.drv-5/dmd which is the same path localDMD wants to be written to. :-)

I don't see how you reached that conclusion. I don't see how the location of the dmd source code is in any way relevant here.

The problem here is that both `rdmd_test` and `Nix` do not do a sufficiently good job at namespacing their temporary files in `$TMP`. The collision could be resolved from etiher side, and you seem to have chosen the D side, but I would suggest to also raise an issue in Nix's issue tracker.

Reply:
Oh, I think I understand what you meant now regarding the DMD dir:

https://github.com/dlang/tools/blob/53735e92398d68eaa2bfd935dd80802fb7f8bc45/posix.mak#L13

It's cloned automatically by the Makefile. I don't know much about Nix packaging, but if you care about the package being buildable into the future or reproducible builds, you will probably want to provide the contents of the `dmd` repository separately, and perhaps specify the path to it using the `DMD_DIR` makefile parameter.

Reply:
> I don't know much about Nix packaging, but if you care about the package being buildable into the future or reproducible builds, you will probably want to provide the contents of the dmd repository separately, and perhaps specify the path to it using the DMD_DIR makefile parameter.

While I was preparing https://github.com/dlang/tools/pull/287 I realized that I already can set DMD_DIR and move it inside the tools directory myself.
So it's no real issue for me anymore.
The remaining question is if this should be the default behavior but should probably better be discussed in the PR.